### PR TITLE
tsxified DateTimePicker + refactored objectifyDates

### LINKED ISSUE #4618: DatePicker poor performance
https://github.com/TerriaJS/terriajs/pull/4576

`objectifyDates` was taking close to 1 second to compute, and was being re-computed way too often.

### PR DESCRIPTION
**Pull request by nf-s**
### tsxified DateTimePicker + moved objectifyDates to DiscretelyTimeVaryingMixin

This is an attempt to fix performance issues related to `objectifyDates` in `DiscretelyTimeVaryingMixin`, which creates a huge object of this structure:

```ts
type DatesObject<T> = {
  [key: number]: T;
  dates: Date[];
  indice: number[];
};
type ObjectifiedDates = DatesObject<ObjectifiedYears>;
type ObjectifiedYears = DatesObject<ObjectifiedMonths>;
type ObjectifiedMonths = DatesObject<ObjectifiedDays>;
type ObjectifiedDays = DatesObject<ObjectifiedHours>;
type ObjectifiedHours = DatesObject<Date[]>;
```

**What I have done**

- Tsxified `DateTimePicker` + turn into StyledComponents + remove all scss
- Moved `objectifyDates` to `DiscretelyTimeVaryingMixin` as a `@computed` property.
- Crunched `objectifyDates` to a single for loop - which has made it much faster

**Problems**

- If the Time Filter is enabled, picking a new `currentDate` re-computes  `objectifyDates`
  - This has something to do with `TimeFilterMixin.timeFilterPropertyName` and `DiscretelyTimeVarying.currentTime`

**Further work**

- Do we need to objectify all dates at once? Can we do it as the user interacts with the DatePicker?
- Update `react-datepicker`

**Test using**

Use `Data-Blended` layer

- Old datepicker = http://ci.terria.io/next/#clean&https://gist.githubusercontent.com/nf-s/3d77d4acb9105283226444677e011dd9/raw/290e89b66e2b2614462b33efb3628dbdd0ac7b99/wms-multidim.json

- New datepicker = http://ci.terria.io/fix-datepicker/#clean&https://gist.githubusercontent.com/nf-s/3d77d4acb9105283226444677e011dd9/raw/290e89b66e2b2614462b33efb3628dbdd0ac7b99/wms-multidim.json

### Checklist

-  [x] Add tests
-  [x] ~Change scss vars to use theme vars~ Everything is styled components!
-  [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Comment by soyarsauce** [comment]
Double check if this addresses https://github.com/TerriaJS/terriajs/issues/4518

**Comment by nf-s** [comment]
> Double check if this addresses #4518

This is fixed now.

**Review by rowanwins — APPROVED** [review]
Looks good to me from the refactored objectifyDates perspective. Can't comment much on the tsx or style components but it works :)

File: lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+            width: 30px;
+            height: 30px;
+          `}
+          onClick={event => {
+            event.stopPropagation();
+          }}
+        >
+          {this.props.showCalendarButton && (
+            <button
+              css={`
```
Comment (soyarsauce) [inline lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx]: Add some sort of title - only especially important as the button doesn't have any text in it (it's just an icon as sole child)

```suggestion
            <button
              title={t("diffTool.datePicker.dateButtonTitle")}
              css={`
```

File: lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+                  ? `
+                  top: 40px;
+                  left: -190px;
+                `
+                  : ""}
+              `}
+              className={"scrollbars"}
+            >
+              <BackButton
+                css={`
```
Comment (soyarsauce) [inline lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx]: Add an appropriate title here too

**Review by soyarsauce — CHANGES_REQUESTED** [review]
Looks good, great work - and understanding the styles are just ripped straight from the old scss ones, it's a good start. Should aim to use the new generic components wherever possible, let's just go for theme for now.

File: lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+  span {
+    display: inline-block;
+    height: 10px;
+    width: 2px;
+    margin-top: 1px;
+    margin-right: ${p => p.marginRight}px;
+  }
+`;
+
+const Grid = styled.div`
```
[the remark this answers is not in the document]
Comment (nf-s) [reply lib/ReactViews/BottomDock/Timeline/DateTimePicker.tsx:77]: alright i'll leave this for now

**Review by soyarsauce — APPROVED** [review]

**Pull request merged**